client/asset/eth/eth.go

```diff
@@ -1768,16 +1778,32 @@ func (w *ETHWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin, uint6
 		swapVal += contract.Value
 	}
 
-	oneSwap, _, _, err := w.swapGas(1, swaps.Version)
+	// Set the gas limit as high as reserves will allow.
+	n := len(swaps.Contracts)
+	oneSwap, nSwap, _, err := w.swapGas(n, swaps.Version)
 	if err != nil {
 		return fail("error getting gas fees: %v", err)
 	}
-
-	gasLimit := oneSwap * uint64(len(swaps.Contracts))
+	gasLimit := oneSwap * uint64(n) // naive unbatched, higher but not realistic
 	fees := gasLimit * swaps.FeeRate
-
 	if swapVal+fees > reservedVal {
-		return fail("unfunded swap: %d < %d", reservedVal, swapVal+fees)
+		if n == 1 {
+			return fail("unfunded swap: %d < %d", reservedVal, swapVal+fees)
+		}
+		w.log.Warnf("Unexpectedly low reserves for %d swaps: %d < %d", n, reservedVal, swapVal+fees)
+		// Since this is a batch swap, attempt to use the realistic limits.
+		gasLimit = nSwap
+		fees = gasLimit * swaps.FeeRate
```

Comment:
`fees` is a return from `Swap`, so arguably we should be returning this correctly batched fees always.
What does everyone think about how presently we call `w.initiate` below with the naive unbatched gas limit?  Is this a bug or just conservative?  This reserves condition I'm adding here could potentially be the normal path...

Reply:
Actually the dynamic fees update is quite jarring with batched swap because of this quirk / conservative tx authoring with batched swaps.

Reply:
hmm, considering this more, if we do *not* use the properly batched gas limits, the tx might be rejected if it even approaches the max swaps per tx limit

Reply:
> hmm, considering this more, if we do not use the properly batched gas limits, the tx might be rejected if it even approaches the max swaps per tx limit

The `maxSwapsInTx` is calculated using the individual `Swap` value not the batched value, so I don't think this would happen. The downside of using the batched estimate is that the real value would be more than 10% higher than the estimate, which is unlikely, so might as well go with that if only to just make the dynamic fee update less jarring.

Reply:
Consider if it was a batch of say, _half_ of `maxSwapsInTx`.  Here in `Swap` we call `oneSwap, nSwap, _, err := w.swapGas(n, swaps.Version)`, and assume it gives us the expected hard-coded gas values for both `oneSwap` and `nSwap`.  Why would we use `oneSwap * n` when we know `nSwap` is correct for this batch init tx?
Because we do use `oneSwap * n`, that could go over the `maxTxFeeGwei` limit even though we're far less than `maxSwapsInTx`.  The transaction would not actually consume that gas, but I'm wondering at what level the limit is enforced?  The EVM could easily decide not to even try executing the txn if the gas limit were set too high.

Reply:
Just saw your response.

> The maxSwapsInTx is calculated using the individual Swap value not the batched value, so I don't think this would happen

Why would we do that though?  Shouldn't it be based on the batched value?

Reply:
> Why would we use `oneSwap * n` when we know `nSwap` is correct for this batch init tx?
> Because we do use `oneSwap * n` (to be conservative?), that could go over the `maxTxFeeGwei` limit even though we're far less than `maxSwapsInTx`.

Good point. We used `oneSwap * n` because that was what we planned for in the worst case, but since we are batching here and we better understand the gas limit now, `nSwap` seems appropriate.

Reply:
Asking because we've testing as-is and there are two approves, and I hate to fundamentally change PRs at that stage.

Reply:
@martonp has rightly pointed out that the batch savings is not huge, and because of how `maxSwapsInTx` is computed with the individual Swap value (that's ver questionable too BTW), this is fine as-is.  Perhaps we should look again at this later, perhaps with 0.6.1.

Reply:
Sounds good.